File: dotnet/source/amp.eventing/DefaultEventConsumer.cs
using System;
using System.Collections.Generic;
using amp.messaging;
using cmf.bus;
using cmf.eventing;

namespace amp.eventing
{
    public class DefaultEventConsumer : MessageReceiver, IEventConsumer
    {
        public DefaultEventConsumer(IEnvelopeReceiver envelopeReceiver) 
            : base(envelopeReceiver)
        {
        }

        public DefaultEventConsumer(IEnvelopeReceiver envelopeReceiver, List<IMessageProcessor> processingChain) 
            : base(envelopeReceiver, processingChain)
        {
        }


        public void Subscribe<TEvent>(Action<TEvent, IDictionary<string, string>> handler) where TEvent : class
        {
            base.ReceiveMessage(handler);
        }

        public void Subscribe(IEventHandler handler)
        {
            base.ReceiveMessage(new EventMessageHandler(handler));
        }

        private class EventMessageHandler : IMessageHandler
        {
            private readonly IEventHandler _handler;

            public EventMessageHandler(IEventHandler handler)
            {
                _handler = handler;
            }

            public string Topic
            {
                get { return _handler.Topic; }
            }

            public object Handle(object message, IDictionary<string, string> headers)
            {
                return _handler.Handle(message, headers);
            }

            public object HandleFailed(Envelope env, Exception ex)
            {
                return _handler.HandleFailed(env, ex);
            }
        }
    }
}

File: java/amp-eventing/src/main/java/amp/eventing/DefaultEventConsumer.java
package amp.eventing;


import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import amp.messaging.IMessageHandler;
import amp.messaging.IMessageProcessor;
import amp.messaging.MessageException;
import amp.messaging.MessageReceiver;
import cmf.bus.Envelope;
import cmf.bus.IEnvelopeFilterPredicate;
import cmf.bus.IEnvelopeReceiver;
import cmf.eventing.IEventConsumer;
import cmf.eventing.IEventHandler;


/**
 * Created with IntelliJ IDEA.
 * User: jar349
 * Date: 5/1/13
 */
public class DefaultEventConsumer extends MessageReceiver implements IEventConsumer {

    private static final Logger LOG = LoggerFactory.getLogger(DefaultEventConsumer.class);

        public DefaultEventConsumer(IEnvelopeReceiver envelopeReceiver) {
        super(envelopeReceiver);
    }

    public DefaultEventConsumer(IEnvelopeReceiver envelopeReceiver, List<IMessageProcessor> processorChain) {
        super(envelopeReceiver, processorChain);
    }

    @Override
    public <TEVENT> void subscribe(IEventHandler<TEVENT> handler) throws MessageException, IllegalArgumentException {
        subscribe(handler, new IEnvelopeFilterPredicate() {
			
			@Override
			public boolean filter(Envelope envelope) {
				return true;
			}
		});
    }
    
    @Override
	public <TEVENT> void subscribe(IEventHandler<TEVENT> handler,
			IEnvelopeFilterPredicate predicate) throws MessageException {

    	onMessageReceived(new EventMessageHandler<TEVENT>(handler), predicate);
    }
    
    private static class EventMessageHandler<TEVENT> implements IMessageHandler<TEVENT>{
    	private static final Logger LOG = LoggerFactory.getLogger(EventMessageHandler.class);

        private final IEventHandler<TEVENT> _handler;
  
        public EventMessageHandler(IEventHandler<TEVENT> handler){
            _handler = handler;
        }

        public Class<TEVENT> getMessageType(){
            return _handler.getEventType(); 
        }

        @Override
		public Object handle(TEVENT message, Map<String, String> headers){
            _handler.handle(message, headers);
            return null;
        }

        @Override
        public Object handleFailed(Envelope env, Exception ex){
        	LOG.warn("Failed to handle envelope.", ex);
            return null;
        }
    }
}
